metronome/checkpointing/service/test/src/io/iohk/metronome/checkpointing/service/CheckpointingServiceProps.scala

```diff
@@ -0,0 +1,407 @@
+          .sequence
+
+        // repeated execution must fail because we're trying to execute a block of lower height
+        // than the last executed block
+        execution >>
+          execution
+            .onErrorHandle(_.isInstanceOf[IllegalStateException])
+            .map(_ == true)
+      }
+    }
+
+  //TODO: Validate transactions PM-3131/3132
+  //      use a mocked interpreter client that always evaluates blocks as valid
+  property("parallel validation") = forAll(TestFixture.gen(minChain = 4)) {
+    fixture =>
+      run(fixture) { res =>
+        import fixture._
+        import res._
+
+        // validation in random order so blocks need to be persisted first
+        val persistBlocks = store.runReadWrite {
+          batch.map(blockStorage.put).sequence
+        }
+
+        def validation(
+            validating: Ref[Task, Boolean],
+            achievedPar: Ref[Task, Boolean]
+        ) =
+          Task.parSequence {
+            Random
+              .shuffle(batch)
+              .map(b =>
+                for {
+                  v <- validating.getAndSet(true)
+                  _ <- achievedPar.update(_ || v)
+                  r <- checkpointingService.validateBlock(b)
+                  _ <- validating.set(false)
+                } yield r.getOrElse(false)
+              )
+          }
+
+        for {
+          _ <- persistBlocks
+
+          // used to make sure that parallelism was achieved
+          validating  <- Ref[Task].of(false)
+          achievedPar <- Ref[Task].of(false)
+
+          result     <- validation(validating, achievedPar)
+          par        <- achievedPar.get
+          ledgerTree <- ledgerTreeRef.get
+        } yield {
+          val ledgerTreeUpdated = batch.forall(b => ledgerTree.contains(b.hash))
+
+          all(
+            "parallelism achieved" |: par,
+            "validation successful" |: result.forall(identity),
+            "ledgerTree updated" |: ledgerTreeUpdated
+          )
+        }
+      }
+  }
+
+  //TODO: Validate transactions PM-3131/3132
+  //      use a mocked interpreter client that always evaluates blocks as valid
+  property("execution parallel to validation") = forAllNoShrink {
+    for {
+      f   <- TestFixture.gen(minChain = 4)
+      ext <- TestFixture.genBlockChain(f.batch.last, f.finalLedger)
+    } yield (f, f.batch ++ ext)
+  } { case (fixture, validationBatch) =>
+    run(fixture) { res =>
+      import fixture._
+      import res._
+
+      // validation in random order so blocks need to be persisted first
+      val persistBlocks = store.runReadWrite {
+        validationBatch.map(blockStorage.put).sequence
+      }
+
+      def validation(
+          validating: Ref[Task, Boolean],
+          executing: Ref[Task, Boolean],
+          achievedPar: Ref[Task, Boolean]
+      ) =
+        Random
+          .shuffle(validationBatch)
+          .map(b =>
+            Task.defer {
```

Comment:
What's the extra `defer` for?

Reply:
When I initially checked the parallelism by adding simple `println`s (within the `flatMap` bodies inside the methods) I found out that all validation are run before execution, as if the Tasks were actually being run upon creation. `Task.defer` seemed to rectify this behaviour. I also added the parallelism checks to make sure.

However now, I am not able to reproduce that behaviour, even when I just do `.map(b => checkpointingService.validateBlock(b)) //or execute`. I wish I could explain this, but all I can do is remove those `Task.defer`.

Regarding the parallelism checks though. We've had a [failure](https://app.circleci.com/pipelines/github/input-output-hk/metronome/471/workflows/e9b3767d-d3e6-4b23-94a6-17b5f02bd286/jobs/527). We've had a [success](https://app.circleci.com/pipelines/github/input-output-hk/metronome/473/workflows/99647d7b-3fc9-4894-b53d-365bcfa475e5/jobs/529), too. On a single occasion, I also observed a failure locally. I guess one way to improve this would be to increase minimum batch sizes (more opportunity for parallelism), but it may be that these tests will always be flaky, especially if running on a single-thread scheduler. What do you think? Should we keep them?

Reply:
Sometimes I put in a short `sleep` to simulate time and raise the chances of things actually happening concurrently. While you could do that in this case it would probably not be inside a critical section, where it could cause any issues. 

What I'd try is to [classify](https://github.com/typelevel/scalacheck/blob/main/doc/UserGuide.md#collecting-generated-test-data) tests by whether they achieved parallelism or not. That would give us the confidence that the test is written correctly, even if occasionally it's not actually concurrent on CI.

Reply:
`classify` is pretty neat, but I would still like to specify a requirement on it, like "require 50% of test cases to be achieve parallelism". Doesn't seem possible though. I added a note to the maintainer to make sure to observe these stats when modifying the test or the service.